internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"sync"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/overlord/servstate"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+type LogManager struct {
+	forwarders []*logForwarder
+	// service -> targets selected from plan
+	planTargets map[string]map[string]*plan.LogTarget
+	// Keep service config to know what changed
+	services map[string]*plan.Service
+	// Keep ring buffer for each running service
+	buffers map[string]*servicelog.RingBuffer
+	// Keep iterator for each (service, target) pair
+	iterators map[string]map[plan.LogTarget]servicelog.Iterator
+	// Mutex for reading/writing manager structs
+	mu sync.Mutex
+}
+
+func NewLogManager() *LogManager {
+	return &LogManager{
+		buffers: map[string]*servicelog.RingBuffer{},
+	}
+}
+
+// PlanChanged is called by the service manager when the plan changes. We stop
+// all running forwarders, and start new forwarders based on the new plan.
+// There are three distinct cases when creating the new forwarders:
+//   - Existing service and target: create the forwarder using the same
+//     iterator, to avoid log duplication.
+//   - Existing service, new target: create the forwarder using a new iterator
+//     over the existing buffer.
+//   - New service: don't create anything yet. We have to wait until
+//     ServiceStarted is called, and we are given a reference to the service's
+//     ringbuffer.
+func (m *LogManager) PlanChanged(pl *plan.Plan) {
```

Comment:
I realize this function is still relatively complicated, but I think that's inherent complexity now. The flow and structure of it is a lot nicer now, thanks!